# 固有値が複素数の場合

ここまで行列$\bf A$の固有値$\lambda$と固有ベクトル$\bf p$を求める方法を学習しました．  
固有値は，固有方程式$\text{det}({\bf A}-\lambda{\bf I})=0$を解くことによって求まります．
この固有方程式は，行列の次数と同じ次数の代数方程式です．
例えば，行列$\bf A$を2次正方行列とすると，固有方程式は2次方程式になります．

これまで幾何学的な意味と一致させるために，スカラーを実数としてきました．
固有値問題の例では，固有値が実数となる行列で解説してきました．
しかし，固有方程式の解が複素数になることもあります．
ここでは，そのようなケースについて説明します．

## 線形変換としての回転

ここでは，次の行列を扱います．
これは，反時計回りに30度回転させる線形変換です．

> ${\bf A} =
\left(\begin{array}{c}
\cos{\frac{\pi}{6}} & -\sin{\frac{\pi}{6}} \\
\sin{\frac{\pi}{6}} &  \cos{\frac{\pi}{6}} \\
\end{array}\right)$
$=
\left(\begin{array}{c}
\frac{\sqrt{3}}{2} & -\frac{1}{2} \\
\frac{1}{2} &  \frac{\sqrt{3}}{2} \\
\end{array}\right)$

この行列について，固有値を求めていきます．

#### 固有値

まず，固有多項式を計算します．

$$\begin{eqnarray}
\text{det}({\bf A}-\lambda{\bf I}) &=&
\text{det}\left\{
\left(\begin{array}{c}
\frac{\sqrt{3}}{2} & -\frac{1}{2} \\
\frac{1}{2} & \frac{\sqrt{3}}{2} \\
\end{array}\right)
-
\left(\begin{array}{c}
\lambda & 0 \\
0 & \lambda \\
\end{array}\right)
\right\} \\
&=&
\left|\begin{array}{c}
\frac{\sqrt{3}}{2}-\lambda & -\frac{1}{2} \\
\frac{1}{2} & \frac{\sqrt{3}}{2}-\lambda \\
\end{array}\right| \\
&=&
\left(\frac{\sqrt{3}}{2}-\lambda\right)^2 +\left(\frac{1}{2}\right)^2 \\
&=&
\frac{3}{4}-\sqrt{3}\lambda+{\lambda}^2+\frac{1}{4} \\
&=&
{\lambda}^2-\sqrt{3}\lambda+1 \\
\end{eqnarray}$$

よって，固有方程式
${\lambda}^2-\sqrt{3}\lambda+1=0$
を解くと，

> $\displaystyle \lambda = \frac{\sqrt{3}}{2} \pm \frac{1}{2}i$

となります．
この行列の固有値は複素数です．

#### 固有ベクトル

固有値$\lambda_1 = \frac{\sqrt{3}}{2}+\frac{1}{2}i$ に対する固有ベクトルを計算します．

$${\bf A}-\lambda_1{\bf I}
=
\left(\begin{array}{c}
\frac{\sqrt{3}}{2}-\lambda_1 & -\frac{1}{2} \\
\frac{1}{2} & \frac{\sqrt{3}}{2}-\lambda_1 \\
\end{array}\right) 
=
\left(\begin{array}{c}
-\frac{1}{2}i & -\frac{1}{2} \\
\frac{1}{2} & -\frac{1}{2}i \\
\end{array}\right) $$

となるので，
$({\bf A}-\lambda_1{\bf I}){\bf p}={\bf 0}$
の中身を連立方程式で書くと，

$$\left\{ \begin{eqnarray} 
ix + y &=& 0 \\
x - iy &=& 0 \\
\end{eqnarray} \right. $$

となります．
ここで，$x=c_1$と置くと，$y=-ic_1$となるので固有ベクトルは，

> ${\bf p}_1 = c_1 \left(\begin{array}{c} 1 \\ -i \\ \end{array}\right)$

となります．

固有値$\lambda_2 = \frac{\sqrt{3}}{2}-\frac{1}{2}i$ に対する固有ベクトルも同様に計算します．

$${\bf A}-\lambda_2{\bf I}
=
\left(\begin{array}{c}
\frac{\sqrt{3}}{2}-\lambda_2 & -\frac{1}{2} \\
\frac{1}{2} & \frac{\sqrt{3}}{2}-\lambda_2 \\
\end{array}\right) 
=
\left(\begin{array}{c}
\frac{1}{2}i & -\frac{1}{2} \\
\frac{1}{2} & \frac{1}{2}i \\
\end{array}\right) $$

となるので，
$({\bf A}-\lambda_2{\bf I}){\bf p}={\bf 0}$
の中身を連立方程式で書くと，

$$\left\{ \begin{eqnarray} 
ix - y &=& 0 \\
x + iy &=& 0 \\
\end{eqnarray} \right. $$

となります．
ここで，$x=c_2$と置くと，$y=ic_2$となるので固有ベクトルは，

> ${\bf p}_2 = c_2 \left(\begin{array}{c} 1 \\ i \\ \end{array}\right)$

となります．

## 幾何学的解釈

さて，行列$\bf A$の固有値と固有ベクトルが求まりましたので，その幾何学的解釈を与えます．
元々の空間が2次元ユークリッド空間，つまり2次元実数ベクトル空間でした．
しかし，求まった固有ベクトルは複素数ベクトルになってしまいました．

これまで見てきた実数の固有値と固有ベクトルの場合，行列$\bf A$によって引き起こされる線形変換を固有ベクトルに作用させると，
ベクトルの方向を維持して大きさが固有値倍になりました．
固有値が負の実数の場合は，ベクトルの向きが逆になります．

しかし，固有値が複素数になると，この説明はナンセンスなものとなっています．
そこで，行列$\bf A$によるベクトル場の画像を確認してみましょう．

![回転の線形写像](./images/vecfield_rotation.jpg)

このように想定通り反時計方向に30度の回転になっています．

まず，固有値ですが，

> $\displaystyle \lambda_1 = \frac{\sqrt{3}}{2}+\frac{1}{2}i$

> $\displaystyle \lambda_2 = \frac{\sqrt{3}}{2}-\frac{1}{2}i$

これらは，お互いに複素共役になっています．
つまり，$\overline{\lambda_1} = \lambda_2$となっています．

ここで固有値の大きさである絶対値を計算します．
ただし，複素数の絶対値はその数とその複素共役を掛けた値になります．

$$ \begin{eqnarray}
{|{\lambda_1}|}^2 &=&
\overline{\lambda_1}{\cdot}\lambda_1 \\
&=&
\lambda_2{\cdot}\lambda_1 \\
&=&
\left(\frac{\sqrt{3}}{2}-\frac{1}{2}i\right)\left(\frac{\sqrt{3}}{2}+\frac{1}{2}i\right) \\
&=&
\frac{\sqrt{3}}{2}{\cdot}\frac{\sqrt{3}}{2} + \frac{\sqrt{3}}{2}{\cdot}\frac{1}{2}i - \frac{\sqrt{3}}{2}{\cdot}\frac{1}{2}i - \frac{1}{2}i{\cdot}\frac{1}{2}i \\
&=&
\frac{3}{4}+\frac{1}{4} \\
&=& 1 \\
\end{eqnarray}$$

これが意味することは，ベクトルの変換前と変換後の大きさが変わらないということです．
確かに，単純な回転ですので，ベクトルの大きさは変わっておりません．

もし，固有値が複素数で，かつ，ベクトルの大きさが1より大きい場合は，回転しながら伸長しています．
また，複素数である固有値の大きさが1より小さければ，回転しながら縮小します．

*****
## Pythonによる確認

それでは，この固有値問題についてもPythonで確認しましょう．

まずはNumPyライブラリーをインポートします．

In [1]:
import numpy as np

行列${\bf A} =
\left(\begin{array}{c}
\frac{\sqrt{3}}{2} & -\frac{1}{2} \\
\frac{1}{2} &  \frac{\sqrt{3}}{2} \\
\end{array}\right)$
を配列オブジェクトとして定義します．

In [2]:
A = np.array([[np.sqrt(3)/2,-1/2],[1/2,np.sqrt(3)/2]])
A

array([[ 0.8660254, -0.5      ],
       [ 0.5      ,  0.8660254]])

<font color=green>numpy.linalg.eig()</font>関数によって，固有値と固有ベクトルを求めます．

> eigen_values,eigen_vectors = np.linalg.eig(A)

結果は，固有値と固有ベクトルが配列に格納されて返されます．

In [3]:
eigen_values,eigen_vectors = np.linalg.eig(A)

固有値を確認します．

In [4]:
eigen_values

array([ 0.8660254+0.5j,  0.8660254-0.5j])

Pythonでは虚数単位をアルファベットの **j** で表します．
$\frac{\sqrt{3}}{2} \approx 0.8660254$なので，固有値として，

> $\displaystyle \lambda_1 = \frac{\sqrt{3}}{2}+\frac{1}{2}i$

> $\displaystyle \lambda_2 = \frac{\sqrt{3}}{2}-\frac{1}{2}i$

が求まりました．

$\lambda_1$に対応する固有ベクトルである eigen_vectors[:,0] を確認します．

In [5]:
eigen_vectors[:,0]

array([ 0.00000000-0.70710678j, -0.70710678+0.j        ])

$\frac{\sqrt{2}}{2} \approx 0.70710678$なので，この正規化された固有ベクトルは，

> ${\bf p}_1 = \left(\begin{array}{c} -\frac{\sqrt{2}}{2}i \\ -\frac{\sqrt{2}}{2} \\ \end{array}\right) $
$\propto \left(\begin{array}{c} 1 \\ -i \\ \end{array}\right)$

となっていっます．

また，$\lambda_2$に対応する固有ベクトルである eigen_vectors[:,1] を確認します．

In [6]:
eigen_vectors[:,1]

array([ 0.00000000+0.70710678j, -0.70710678-0.j        ])

この正規化された固有ベクトルは，

> ${\bf p}_2 = \left(\begin{array}{c} \frac{\sqrt{2}}{2}i \\ -\frac{\sqrt{2}}{2} \\ \end{array}\right) $
$\propto \left(\begin{array}{c} 1 \\ i \\ \end{array}\right)$

です．

${\bf p}_1$，${\bf p}_2$の両方とも手計算で求めた固有ベクトルと同じ固有空間を代表するベクトルになっています．

*****
以上のように，固有値が複素数の場合も実数値の場合と同様に解けることが分かりました．